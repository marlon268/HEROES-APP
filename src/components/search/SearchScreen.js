import React, { useMemo } from 'react'
import queryString from "query-string";

import { useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

   // npm install query-string; lo instalamos para obtener mas facilmente la query

   const location = useLocation();

   const { q = "" } = queryString.parse(location.search)


   const [valuesForm, handleInputChange] = useForm({
      searchText: q
   })

   const { searchText } = valuesForm;

   const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

   const handleSearch = (e) => {

      e.preventDefault()

      history.push(`?q=${searchText}`)
   }

   return (
      <div>
         <h1>Search</h1>
         <hr />

         <div>
            <div>
               <h4> Search Form </h4>
               <hr />
               <form onSubmit={handleSearch}>
                  <input
                     autoComplete="off"
                     name="searchText"
                     type="text"
                     placeholder="Find your hero"
                     className="form-control"
                     value={searchText}
                     onChange={handleInputChange}
                  />

                  <button
                     type="submit"
                     className="brn m-1"
                  >
                     Search...
                  </button>
               </form>
               <h4>Results</h4>
               <hr /></div>
            <div className="container1">

               {heroesFiltered.length < 1 ? <h1 className="text-danger">Heroe no existe</h1> :
                  heroesFiltered.map((hero) => (
                     <HeroCard
                        key={hero.id}
                        {...hero}
                     />
                  ))
               }
            </div>
         </div>
      </div>
   )
}
