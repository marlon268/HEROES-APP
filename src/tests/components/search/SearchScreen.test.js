import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router-dom'
import { SearchScreen } from '../../../components/search/SearchScreen'

describe('Pruebas en <SearchScreen />', () => {

   test('Debe de mostrarse correctamente con valores por defecto', () => {

      const wrapper = mount(
         <MemoryRouter initialEntries={['/search']}>
            <Route path="/search" component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper).toMatchSnapshot()
      expect(wrapper.find('HeroCard').exists()).toBe(true)

   });

   test('Debe de mostrar a batman y el input con el valor del queryString', () => {

      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=batman']}>
            <Route path="/search" component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper.find('input').prop('value')).toBe('batman')
      expect(wrapper).toMatchSnapshot()

   });

   test('Debe de mostrar un error si no se encuentra el hero', () => {

      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=batman123']}>
            <Route path="/search" component={SearchScreen} />
         </MemoryRouter>
      );

      expect(wrapper.find('.text-danger').exists()).toBe(true)
      expect(wrapper).toMatchSnapshot()

   });

   test('Debe de llamar el push del history', () => {

      const history = {
         push: jest.fn()
      };

      const wrapper = mount(
         <MemoryRouter initialEntries={['/search?q=batman']}>
            <Route path="/search" component={(porps) => <SearchScreen history={history} />} />
         </MemoryRouter>
      );

      wrapper.find('input').simulate('change', {
         target: {
            name: 'searchText',
            value: 'batman'
         }
      });

      wrapper.find('form').prop('onSubmit')({
         preventDefault() { }
      });

      expect(history.push).toHaveBeenCalledWith(`?q=batman`)

   })



})
