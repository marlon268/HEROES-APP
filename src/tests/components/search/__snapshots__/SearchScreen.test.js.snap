// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar a batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div>
    <div>
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="brn m-1"
          type="submit"
        >
          Search...
        </button>
      </form>
      <h4>
        Results
      </h4>
      <hr />
    </div>
    <div
      className="container1"
    >
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-batman"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Batman
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Batman"
            className="img"
            src=""
          />
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrar un error si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div>
    <div>
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman123"
        />
        <button
          className="brn m-1"
          type="submit"
        >
          Search...
        </button>
      </form>
      <h4>
        Results
      </h4>
      <hr />
    </div>
    <div
      className="container1"
    >
      <h1
        className="text-danger"
      >
        Heroe no existe
      </h1>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div>
    <div>
      <h4>
         Search Form 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="brn m-1"
          type="submit"
        >
          Search...
        </button>
      </form>
      <h4>
        Results
      </h4>
      <hr />
    </div>
    <div
      className="container1"
    >
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-batman"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Batman
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Batman"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-superman"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Superman
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Superman"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-flash"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Flash
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Flash"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-green"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Green Lantern
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Green Lantern"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-arrow"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Green Arrow
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Green Arrow"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-wonder"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Wonder Woman
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Wonder Woman"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-martian"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Martian Manhunter
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Martian Manhunter"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-robin"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Robin/Nightwing
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Robin/Nightwing"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-blue"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Blue Beetle
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Blue Beetle"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/dc-black"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Black Canary
          </h5>
          <h5
            className="absolute publisher"
          >
            DC Comics
          </h5>
          <img
            alt="Black Canary"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-spider"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Spider Man
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Spider Man"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-captain"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Captain America
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Captain America"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-iron"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Iron Man
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Iron Man"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-thor"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Thor
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Thor"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-hulk"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Hulk
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Hulk"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-wolverine"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Wolverine
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Wolverine"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-daredevil"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Daredevil
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Daredevil"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-hawkeye"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Hawkeye
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Hawkeye"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-cyclops"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Cyclops
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Cyclops"
            className="img"
            src=""
          />
        </a>
      </div>
      <div
        className="container-child"
      >
        <a
          href="/hero/marvel-silver"
          onClick={[Function]}
        >
          <h5
            className="absolute superhero"
          >
            Silver Surfer
          </h5>
          <h5
            className="absolute publisher"
          >
            Marvel Comics
          </h5>
          <img
            alt="Silver Surfer"
            className="img"
            src=""
          />
        </a>
      </div>
    </div>
  </div>
</div>
`;
